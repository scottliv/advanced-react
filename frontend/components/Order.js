import React, { Component } from "react";
import PropTypes from "prop-types";
import { Query } from "react-apollo";
import { format } from "date-fns";
import Head from "next/head";
import gql from "graphql-tag";
import formatMoney from "../lib/formatMoney";
import Error from "./ErrorMessage";
import OrderStyles from "./styles/OrderStyles";
import OrderItem from "./OrderItem";

const SINGLE_ORDER_QUERY = gql`
  query SINGLE_ORDER_QUERY($id: ID!) {
    order(id: $id) {
      id
      charge
      total
      createdAt
      user {
        id
      }
      items {
        id
        title
        description
        price
        quantity
        image
      }
    }
  }
`;

class Order extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired
  };
  render() {
    return (
      <Query query={SINGLE_ORDER_QUERY} variables={{ id: this.props.id }}>
        {(data, error, loading) => {
          if (error) return <Error error={error} />;
          if (loading) return <p>Loading...</p>;
          const order = data.data.order;
          console.log(order);
          return (
            <OrderStyles>
              <Head>
                <title>Sick Fits Order: {order && order.id}</title>
              </Head>
              <p>
                <span>Order:</span>
                <span>{this.props.id}</span>
              </p>
              <p>
                <span>Charge</span>
                <span>{order.charge}</span>
              </p>

              <p>
                <span>Date</span>
                <span>
                  {order && format(order.createdAt, "MMMM d, YYYY h:mm a")}
                </span>
              </p>
              <p>
                <span>Total</span>
                <span>{order && formatMoney(order.total)}</span>
              </p>
              <p>
                <span>Item Count:</span>
                <span>{order.items.length}</span>
              </p>
              <div className="items">
                {order.items.map(item => (
                  <OrderItem item={item} key={item.id} />
                ))}
              </div>
            </OrderStyles>
          );
        }}
      </Query>
    );
  }
}

export default Order;
