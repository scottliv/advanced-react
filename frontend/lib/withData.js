import withApollo from "next-with-apollo";
import ApolloClient from "apollo-boost";
import { endpoint, endpointProd } from "../config";
import { LOCAL_STATE_QUERY } from "../components/Cart";

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === "development" ? endpoint : endpointProd,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: "include"
        },
        headers
      });
    },
    // local data
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            //We need a graphql query to read from cache
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY
            });
            const data = {
              data: { cartOpen: !cartOpen }
            };
            cache.writeData(data);
            return data;
          }
        }
      },
      defaults: {
        cartOpen: false
      }
    }
  });
}

export default withApollo(createClient);
