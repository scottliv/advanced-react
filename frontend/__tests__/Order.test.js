import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import { MockedProvider } from "react-apollo/test-utils";
import wait from "waait";
import Order, { SINGLE_ORDER_QUERY } from "../components/Order";
import { fakeOrder } from "../lib/testUtils";

const mocks = [
  {
    request: {
      query: SINGLE_ORDER_QUERY,
      variables: { id: "ord123" }
    },
    result: {
      data: {
        order: { ...fakeOrder() }
      }
    }
  }
];

describe("<Order />", () => {
  it("renders and matches snapshot", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Order id="ord123" />
      </MockedProvider>
    );

    await wait();
    wrapper.update();

    expect(toJson(wrapper.find("div[data-test='order']"))).toMatchSnapshot();
  });
});
