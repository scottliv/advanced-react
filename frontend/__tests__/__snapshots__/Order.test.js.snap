// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order /> renders and matches snapshot 1`] = `
<div
  className="OrderStyles-sc-4oqalm-0 camMij"
  data-test="order"
>
  <SideEffect(Head)>
    <Head />
  </SideEffect(Head)>
  <p>
    <span>
      Order:
    </span>
    <span>
      ord123
    </span>
  </p>
  <p>
    <span>
      Charge
    </span>
    <span>
      ch_123
    </span>
  </p>
  <p>
    <span>
      Date
    </span>
    <span>
      March 31, 2018 5:00 PM
    </span>
  </p>
  <p>
    <span>
      Total
    </span>
    <span>
      $400
    </span>
  </p>
  <p>
    <span>
      Item Count:
    </span>
    <span>
      2
    </span>
  </p>
  <div
    className="items"
  >
    <OrderItem
      item={
        Object {
          "__typename": "OrderItem",
          "description": "et modi tenetur amet modi reprehenderit omnis",
          "id": "3a430a73-c9e9-4bcc-b46a-41618965ffea",
          "image": "reprehenderit.jpg",
          "price": 4234,
          "quantity": 1,
          "title": "soluta non omnis consequatur enim quia autem",
          Symbol(id): "OrderItem:3a430a73-c9e9-4bcc-b46a-41618965ffea",
        }
      }
      key="3a430a73-c9e9-4bcc-b46a-41618965ffea"
    >
      <div
        className="order-item"
      >
        <img
          alt="soluta non omnis consequatur enim quia autem"
          src="reprehenderit.jpg"
        />
        <div
          className="item-details"
        >
          <h2>
            soluta non omnis consequatur enim quia autem
          </h2>
          <p>
            QTY: 
            1
          </p>
          <p>
            Each: 
            $42.34
          </p>
          <p>
            Subtotal: 
            $42.34
          </p>
          <p>
            et modi tenetur amet modi reprehenderit omnis
          </p>
        </div>
      </div>
    </OrderItem>
    <OrderItem
      item={
        Object {
          "__typename": "OrderItem",
          "description": "sapiente laudantium molestias assumenda quasi adipisci mollitia",
          "id": "2393e090-592f-4d03-b808-c9d81098deec",
          "image": "sint.jpg",
          "price": 4234,
          "quantity": 1,
          "title": "quia sed exercitationem omnis laborum exercitationem est",
          Symbol(id): "OrderItem:2393e090-592f-4d03-b808-c9d81098deec",
        }
      }
      key="2393e090-592f-4d03-b808-c9d81098deec"
    >
      <div
        className="order-item"
      >
        <img
          alt="quia sed exercitationem omnis laborum exercitationem est"
          src="sint.jpg"
        />
        <div
          className="item-details"
        >
          <h2>
            quia sed exercitationem omnis laborum exercitationem est
          </h2>
          <p>
            QTY: 
            1
          </p>
          <p>
            Each: 
            $42.34
          </p>
          <p>
            Subtotal: 
            $42.34
          </p>
          <p>
            sapiente laudantium molestias assumenda quasi adipisci mollitia
          </p>
        </div>
      </div>
    </OrderItem>
  </div>
</div>
`;
